crumbles/src/lib.rs

```diff
@@ -0,0 +1,653 @@
+
+            self.last_snapshot_mut().insert(page_num, clean_page);
+        }
+
+        Ok(())
+    }
+
+    unsafe fn snap(&mut self) -> io::Result<()> {
+        if libc::mprotect(self.bytes.as_mut_ptr().cast(), PAGE_SIZE, PROT_READ)
+            != 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+
+        self.snapshots.push(Snapshot::new());
+
+        Ok(())
+    }
+
+    unsafe fn apply(&mut self) -> io::Result<()> {
+        if libc::mprotect(self.bytes.as_mut_ptr().cast(), PAGE_SIZE, PROT_READ)
+            != 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+
+        let popped_snapshot = self.snapshots.pop().unwrap();
+        if self.snapshots.is_empty() {
+            self.snapshots.push(Snapshot::new());
+        }
+        let snapshot = self.last_snapshot_mut();
+
+        for (page_num, clean_page) in popped_snapshot {
+            snapshot.entry(page_num).or_insert(clean_page);
+        }
+
+        Ok(())
+    }
+
+    unsafe fn revert(&mut self) -> io::Result<()> {
+        let popped_snapshot = self.snapshots.pop().unwrap();
+        if self.snapshots.is_empty() {
+            self.snapshots.push(Snapshot::new());
+        }
+
+        for (page_num, clean_page) in popped_snapshot {
+            let page_offset = page_num * PAGE_SIZE;
+            self.bytes[page_offset..][..PAGE_SIZE]
+                .copy_from_slice(&clean_page[..]);
+        }
+
+        if libc::mprotect(self.bytes.as_mut_ptr().cast(), PAGE_SIZE, PROT_READ)
+            != 0
+        {
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(())
+    }
+
+    fn last_snapshot(&self) -> &Snapshot {
+        self.snapshots.last().unwrap()
+    }
+
+    fn last_snapshot_mut(&mut self) -> &mut Snapshot {
+        self.snapshots.last_mut().unwrap()
+    }
+}
+
+impl Drop for MmapInner {
+    fn drop(&mut self) {
+        unsafe {
+            libc::munmap(self.bytes.as_mut_ptr().cast(), MEM_SIZE);
+        }
+    }
+}
+
+static SIGNAL_HANDLER: Once = Once::new();
+
+// Sets up [`segfault_handler`] to handle SIGSEGV, and returns the previous
+// action used to handle it, if any.
+unsafe fn setup_action() -> sigaction {
+    static OLD_ACTION: OnceLock<sigaction> = OnceLock::new();
+
+    SIGNAL_HANDLER.call_once(|| {
+        let mut sa_mask = MaybeUninit::<sigset_t>::uninit();
+        sigemptyset(sa_mask.as_mut_ptr());
+
+        let act = sigaction {
+            sa_sigaction: segfault_handler as _,
+            sa_mask: sa_mask.assume_init(),
+            sa_flags: SA_SIGINFO,
+            #[cfg(target_os = "linux")]
+            sa_restorer: None,
+        };
+        let mut old_act = MaybeUninit::<sigaction>::uninit();
+
+        if libc::sigaction(libc::SIGSEGV, &act, old_act.as_mut_ptr()) != 0 {
+            process::exit(1);
+        }
+
+        // On Apple Silicon for some reason SIGBUS is thrown instead of SIGSEGV.
+        // TODO should investigate properly
```

Comment:
Is this TODO supposed to be a reminder for the future or did you plan to investigate further before releasing?

Reply:
It's meant to be just a reminder. If a problem occurs on Apple Silicon, we know where to look.